src/Microsoft.OData.Core/Json/JsonValueUtilsAsync.cs

```diff
@@ -0,0 +1,491 @@
+        internal static async Task WriteValueAsync(this TextWriter writer, byte value)
+        {
+            Debug.Assert(writer != null, "writer != null");
+
+            await writer.WriteAsync(value.ToString(CultureInfo.InvariantCulture)).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Write an sbyte value.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="value">SByte value to be written.</param>
+        internal static async Task WriteValueAsync(this TextWriter writer, sbyte value)
+        {
+            Debug.Assert(writer != null, "writer != null");
+
+            await writer.WriteAsync(value.ToString(CultureInfo.InvariantCulture)).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Write a string value.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="value">String value to be written.</param>
+        /// <param name="stringEscapeOption">The string escape option.</param>
+        /// <param name="buffer">Char buffer to use for streaming data.</param>
+        /// <param name="arrayPool">Array pool for renting a buffer.</param>
+        internal static async Task WriteValueAsync(this TextWriter writer, string value, ODataStringEscapeOption stringEscapeOption, Ref<char[]> buffer, ICharArrayPool arrayPool = null)
+        {
+            Debug.Assert(writer != null, "writer != null");
+
+            if (value == null)
+            {
+                await writer.WriteAsync(JsonConstants.JsonNullLiteral).ConfigureAwait(false);
+            }
+            else
+            {
+                await writer.WriteEscapedJsonStringAsync(value, stringEscapeOption, buffer, arrayPool).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Write a byte array.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="value">Byte array to be written.</param>
+        /// <param name="buffer">Char buffer to use for streaming data.</param>
+        /// <param name="arrayPool">Array pool for renting a buffer.</param>
+        internal static async Task WriteValueAsync(this TextWriter writer, byte[] value, Ref<char[]> buffer, ICharArrayPool arrayPool = null)
+        {
+            Debug.Assert(writer != null, "writer != null");
+
+            if (value == null)
+            {
+                await writer.WriteAsync(JsonConstants.JsonNullLiteral).ConfigureAwait(false);
+            }
+            else
+            {
+                await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);
+                await writer.WriteBinaryStringAsync(value, buffer, arrayPool).ConfigureAwait(false);
+                await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);

```

Comment:
Is it also possible to write the quotes here as part of the writing the buffer to reduce the async calls?

Reply:
@habbes Same argument as [here](https://github.com/OData/odata.net/pull/2002#discussion_r584965109). We escape special characters when writing the byte array